File: src/cpp/src/detail/face_color.cpp
#include "meshioplusplus/detail/face_color.hpp"

#include <cmath>
#include <stdexcept>
#include <string>
#include <vector>

#include "meshioplusplus/detail/value_io.hpp"

namespace meshioplusplus {
namespace detail {

namespace {

// "no point/cell data array named 'x' (available: a, b)" -- the writers'
// equivalent of data_unknown_key_message, kept local because this one has to
// name both locations at once.
std::string fc_unknown_message(const Mesh& rMesh, const std::string& rName) {
    std::string msg = "meshio++: no point_data or cell_data array named '" + rName + "'";
    std::string avail;
    for (const std::string& n : rMesh.PointDataNames())
        avail += (avail.empty() ? "" : ", ") + ("point:" + n);
    for (const std::string& n : rMesh.CellDataNames())
        avail += (avail.empty() ? "" : ", ") + ("cell:" + n);
    if (avail.empty())
        return msg + " (the mesh carries no data arrays)";
    return msg + " (available: " + avail + ")";
}

// Components = product of the trailing dimensions. A 1-D array is one scalar
// per entry; anything else spreads its remaining axes across the row.
std::size_t fc_num_components(const NDArray& rArray, std::size_t NumEntries) {
    if (NumEntries == 0)
        return 0;
    const std::size_t total = rArray.Size();
    return total / NumEntries;
}

// Reduce one row to a scalar: the requested component, or the magnitude.
//
// The magnitude sums left to right and takes one sqrt at the end -- the Python
// twin does the same, rather than np.linalg.norm, so the two round alike.
double fc_scalarize(const NDArray& rArray, std::size_t Row, std::size_t NumComponents,
                    const std::optional<int>& rComponent) {
    if (NumComponents == 0)
        return std::nan("");
    const std::size_t base = Row * NumComponents;
    if (rComponent.has_value()) {
        const int c = *rComponent;
        if (c < 0 || static_cast<std::size_t>(c) >= NumComponents)
            throw std::invalid_argument("meshio++: component " + std::to_string(c) +
                                        " is out of range for an array with " +
                                        std::to_string(NumComponents) + " component(s)");
        return read_double(rArray, base + static_cast<std::size_t>(c));
    }
    if (NumComponents == 1)
        return read_double(rArray, base);
    double sum = 0.0;
    for (std::size_t k = 0; k < NumComponents; ++k) {
        const double v = read_double(rArray, base + k);
        sum += v * v;
    }
    return std::sqrt(sum);
}

// Flatten a cell_data array to one scalar per global cell of `rMesh`, in the
// same block-major order mSourceCell and "surface:parent_cell" count in.
std::vector<double> fc_flatten_cell_data(const Mesh& rMesh, const std::string& rName,
                                         const std::optional<int>& rComponent) {
    const std::size_t num_blocks = rMesh.NumCellBlocks();
    if (rMesh.CellDataNumBlocks(rName) != num_blocks)
        throw std::invalid_argument("meshio++: cell_data array '" + rName + "' has " +
                                    std::to_string(rMesh.CellDataNumBlocks(rName)) +
                                    " block(s) but the mesh has " + std::to_string(num_blocks));
    std::vector<double> out;
    for (std::size_t b = 0; b < num_blocks; ++b) {
        const std::size_t n = rMesh.Cells(b).NumCells();
        const NDArray& arr = rMesh.CellData(rName, b);
        const std::size_t ncomp = fc_num_components(arr, n);
        for (std::size_t r = 0; r < n; ++r)
            out.push_back(fc_scalarize(arr, r, ncomp, rComponent));
    }
    return out;
}

// The drawn mesh's per-face parent cell in the source mesh, flattened in the
// drawn mesh's own block-major cell order -- so indexing it by a face's
// mSourceCell yields the source cell that owns it. Empty when the drawn mesh
// carries no provenance (it is the source mesh, undrilled).
std::vector<std::int64_t> fc_flatten_parent_ids(const Mesh& rDraw) {
    const std::string key = "surface:parent_cell";
    const std::size_t num_blocks = rDraw.NumCellBlocks();
    if (!rDraw.HasCellData(key) || rDraw.CellDataNumBlocks(key) != num_blocks)
        return {};
    std::vector<std::int64_t> out;
    for (std::size_t b = 0; b < num_blocks; ++b) {
        const std::size_t n = rDraw.Cells(b).NumCells();
        const NDArray& arr = rDraw.CellData(key, b);
        for (std::size_t r = 0; r < n; ++r)
            out.push_back(read_int(arr, r));
    }
    return out;
}

}  // namespace

std::optional<Rgb> FaceColors::Color(std::size_t Face) const {
    if (!mActive || Face >= mValues.size())
        return std::nullopt;
    const double v = mValues[Face];
    if (!std::isfinite(v))
        return std::nullopt;
    return colormap_lookup(mpTable, color_param(v, mVMin, mVMax));
}

double color_param(double v, double vMin, double vMax) {
    // An all-constant array gives vMin == vMax; the middle of the colormap is a
    // better answer than a division by zero, and matches the Python twin.
    if (!(vMax > vMin))
        return 0.5;
    const double t = (v - vMin) / (vMax - vMin);
    if (!(t > 0.0))
        return 0.0;
    if (t > 1.0)
        return 1.0;
    return t;
}

std::vector<ColorFace> color_faces_from_projection(const std::vector<ProjectedFace>& rFaces) {
    std::vector<ColorFace> out;
    out.reserve(rFaces.size());
    for (const ProjectedFace& f : rFaces)
        out.push_back(ColorFace{f.mNodes, f.mNumNodes, f.mSourceCell});
    return out;
}

std::vector<ColorFace> color_faces_flat(const Mesh& rMesh) {
    std::vector<ColorFace> out;
    // Mirrors both flat writers' loop: block-major over the drawable types,
    // then cell-major, with the global cell counter advancing over skipped
    // blocks too (the mSourceCell convention).
    std::int64_t global_cell_base = 0;
    for (const auto cb : rMesh.CellRange()) {
        const std::int64_t block_base = global_cell_base;
        global_cell_base += static_cast<std::int64_t>(cb.NumCells());
        const std::string& type = cb.Type();
        if (type != "line" && type != "triangle" && type != "quad")
            continue;
        const NDArray& conn = cb.Conn();
        const std::size_t ncols = cols(conn);
        const std::size_t n = cb.NumCells();
        for (std::size_t r = 0; r < n; ++r) {
            ColorFace face;
            face.mNodes = {-1, -1, -1, -1};
            face.mNumNodes = static_cast<std::uint8_t>(ncols < 4 ? ncols : 4);
            for (std::size_t k = 0; k < face.mNumNodes; ++k)
                face.mNodes[k] = read_int(conn, r * ncols + k);
            face.mSourceCell = block_base + static_cast<std::int64_t>(r);
            out.push_back(face);
        }
    }
    return out;
}

FaceColors resolve_face_colors(const ColorSpec& rSpec, const Mesh& rSource, const Mesh& rDraw,
                               const std::vector<ColorFace>& rFaces) {
    FaceColors out;
    if (!rSpec.Active())
        return out;

    if (rSpec.mVMin.has_value() && rSpec.mVMax.has_value() && *rSpec.mVMin > *rSpec.mVMax)
        throw std::invalid_argument("meshio++: vmin must not exceed vmax");

    // Resolve the colormap first: an unknown name should fail before any work.
    out.mpTable = colormap_table(rSpec.mCmap);

    const std::string& name = rSpec.mColorBy;
    const bool is_point = rSource.HasPointData(name);
    if (!is_point && !rSource.HasCellData(name))
        throw std::invalid_argument(fc_unknown_message(rSource, name));

    out.mValues.reserve(rFaces.size());

    if (is_point) {
        // Point data lives on the drawn mesh: a boundary extraction carries
        // point_data through its point compaction, so the drawn mesh's node ids
        // index the already-subset array.
        if (!rDraw.HasPointData(name))
            throw std::invalid_argument("meshio++: point_data array '" + name +
                                        "' did not survive boundary extraction");
        const NDArray& arr = rDraw.PointData(name);
        const std::size_t ncomp = fc_num_components(arr, rDraw.NumPoints());
        for (const ColorFace& f : rFaces) {
            // Mean of the corner values, summed in node order then divided
            // once -- the same shape as the projection's depth accumulation,
            // and what the Python twin reproduces (never np.mean, whose
            // pairwise summation rounds differently).
            double sum = 0.0;
            for (std::uint8_t k = 0; k < f.mNumNodes; ++k)
                sum += fc_scalarize(arr, static_cast<std::size_t>(f.mNodes[k]), ncomp,
                                    rSpec.mComponent);
            out.mValues.push_back(f.mNumNodes == 0 ? std::nan("")
                                                   : sum / static_cast<double>(f.mNumNodes));
        }
    } else {
        const std::vector<double> flat = fc_flatten_cell_data(rSource, name, rSpec.mComponent);
        const std::vector<std::int64_t> parents = fc_flatten_parent_ids(rDraw);
        for (const ColorFace& f : rFaces) {
            std::int64_t cell = f.mSourceCell;
            if (!parents.empty()) {
                if (cell < 0 || static_cast<std::size_t>(cell) >= parents.size()) {
                    out.mValues.push_back(std::nan(""));
                    continue;
                }
                cell = parents[static_cast<std::size_t>(cell)];
            }
            if (cell < 0 || static_cast<std::size_t>(cell) >= flat.size())
                out.mValues.push_back(std::nan(""));
            else
                out.mValues.push_back(flat[static_cast<std::size_t>(cell)]);
        }
    }

    // Auto range over the DRAWN faces, not the whole array: the visible figure
    // then spans the whole colorbar. (ParaView's default is the whole array;
    // this is documented as differing.) Non-finite values are excluded, with
    // FiniteStats' seed-on-first-finite rule rather than +/-DBL_MAX sentinels.
    double lo = 0.0;
    double hi = 0.0;
    bool seen = false;
    for (double v : out.mValues) {
        if (!std::isfinite(v))
            continue;
        if (!seen) {
            lo = v;
            hi = v;
            seen = true;
        } else {
            if (v < lo)
                lo = v;
            if (v > hi)
                hi = v;
        }
    }
    out.mVMin = rSpec.mVMin.has_value() ? *rSpec.mVMin : lo;
    out.mVMax = rSpec.mVMax.has_value() ? *rSpec.mVMax : hi;
    if (out.mVMin > out.mVMax)
        throw std::invalid_argument("meshio++: vmin must not exceed vmax");
    out.mActive = true;
    return out;
}

}  // namespace detail
}  // namespace meshioplusplus

File: src/python/meshioplusplus/_facecolor.py
"""Resolve a named data array into one color per drawn face.

The twin of ``src/cpp/include/meshioplusplus/detail/face_color.hpp`` -- see that
header for the resolution rules. The SVG and TikZ writers are pinned
byte-for-byte against the C++ core (``tests/python/test_svg.py``,
``tests/python/test_tikz.py``), so every arithmetic step here mirrors the C++ one
expression for expression: sums run left to right in index order, values are
cast to float before any arithmetic, and no NumPy reduction (``np.mean``,
``np.linalg.norm``, ``np.nanmin``) is used, since their pairwise summation
rounds differently from a sequential loop.
"""

from __future__ import annotations

import math
from typing import NamedTuple

import numpy as np

from ._colormap import colormap_lookup, colormap_table

# Cell types the flat 2D path draws, in the writers' own enumeration order.
_FLAT_TYPES = ("line", "triangle", "quad")


class ColorSpec(NamedTuple):
    """What the writer was asked to color by."""

    color_by: str | None = None
    component: int | None = None
    cmap: str = "viridis"
    vmin: float | None = None
    vmax: float | None = None
    colorbar: bool = False

    @property
    def active(self) -> bool:
        return bool(self.color_by)


class ColorFace(NamedTuple):
    """One drawn face, as the resolver needs to see it."""

    nodes: object  # corner node ids into the drawn mesh
    source_cell: int  # cell index in the drawn mesh, block-major


class FaceColors(NamedTuple):
    """Per-face scalars and the resolved range, ready for lookup."""

    active: bool
    values: list  # one per face; non-finite => nan_color
    vmin: float
    vmax: float
    table: bytes | None

    def color(self, face: int) -> tuple[int, int, int] | None:
        """The color for one face, or ``None`` to use the writer's nan_color."""
        if not self.active:
            return None
        v = self.values[face]
        if not math.isfinite(v):
            return None
        return colormap_lookup(self.table, color_param(v, self.vmin, self.vmax))


INACTIVE = FaceColors(False, [], 0.0, 1.0, None)


def color_param(v: float, vmin: float, vmax: float) -> float:
    """Normalize a value into ``[0, 1]`` over the range, clamping.

    A degenerate range (``vmin == vmax``, which an all-constant array produces)
    maps everything to the middle of the colormap rather than dividing by zero.
    """
    if not (vmax > vmin):
        return 0.5
    t = (v - vmin) / (vmax - vmin)
    if not (t > 0.0):
        return 0.0
    if t > 1.0:
        return 1.0
    return t


def _num_components(array, num_entries: int) -> int:
    """Components = product of the trailing dimensions."""
    if num_entries == 0:
        return 0
    return int(np.asarray(array).size // num_entries)


def _scalarize(flat, row: int, ncomp: int, component: int | None) -> float:
    """Reduce one row to a scalar: the requested component, or the magnitude.

    The magnitude sums left to right and takes one sqrt at the end -- never
    ``np.linalg.norm`` -- so it rounds exactly as the C++ core does.
    """
    if ncomp == 0:
        return float("nan")
    base = row * ncomp
    if component is not None:
        if component < 0 or component >= ncomp:
            raise ValueError(
                f"meshio++: component {component} is out of range for an array "
                f"with {ncomp} component(s)"
            )
        return float(flat[base + component])
    if ncomp == 1:
        return float(flat[base])
    total = 0.0
    for k in range(ncomp):
        v = float(flat[base + k])
        total += v * v
    return math.sqrt(total)


def _flatten_cell_data(mesh, name: str, component: int | None) -> list:
    """One scalar per global cell, block-major -- the mSourceCell order."""
    blocks = mesh.cell_data[name]
    if len(blocks) != len(mesh.cells):
        raise ValueError(
            f"meshio++: cell_data array {name!r} has {len(blocks)} block(s) but "
            f"the mesh has {len(mesh.cells)}"
        )
    out = []
    for block, values in zip(mesh.cells, blocks):
        n = len(block.data)
        arr = np.asarray(values)
        ncomp = _num_components(arr, n)
        flat = arr.reshape(-1)
        for r in range(n):
            out.append(_scalarize(flat, r, ncomp, component))
    return out


def _flatten_parent_ids(mesh) -> list:
    """Per-cell parent ids of the drawn mesh, or ``[]`` when it has none."""
    key = "surface:parent_cell"
    blocks = mesh.cell_data.get(key)
    if blocks is None or len(blocks) != len(mesh.cells):
        return []
    out = []
    for values in blocks:
        out.extend(int(v) for v in np.asarray(values).reshape(-1))
    return out


def faces_from_projection(faces) -> list:
    """The faces a projected surface draws, in emission order."""
    return [ColorFace(nodes, source) for nodes, _, source in faces]


def faces_flat(mesh) -> list:
    """The faces the flat 2D path draws, in its emission order.

    Enumerates ``line``/``triangle``/``quad`` blocks block-major then
    cell-major -- the same rule, and the same order, both flat writers loop
    over -- with the global cell counter advancing over skipped blocks too.
    """
    out = []
    base = 0
    for block in mesh.cells:
        block_base = base
        base += len(block.data)
        if block.type not in _FLAT_TYPES:
            continue
        conn = np.asarray(block.data, dtype=np.int64)
        for r in range(len(conn)):
            out.append(ColorFace(conn[r][:4], block_base + r))
    return out


def resolve_face_colors(spec: ColorSpec, source, draw, faces) -> FaceColors:
    """Resolve ``spec`` into one scalar per face plus the mapped range.

    ``source`` is the writer's input mesh -- the one the array name refers to;
    ``draw`` is the mesh actually being drawn, equal to ``source`` unless a
    boundary was extracted, in which case its ``"surface:parent_cell"`` carries
    the provenance back to ``source``'s cells.
    """
    if not spec.active:
        return INACTIVE

    if spec.vmin is not None and spec.vmax is not None and spec.vmin > spec.vmax:
        raise ValueError("meshio++: vmin must not exceed vmax")

    # Resolve the colormap first: an unknown name should fail before any work.
    table = colormap_table(spec.cmap)

    name = spec.color_by
    is_point = name in source.point_data
    if not is_point and name not in source.cell_data:
        available = [f"point:{n}" for n in sorted(source.point_data)] + [
            f"cell:{n}" for n in sorted(source.cell_data)
        ]
        detail = (
            f" (available: {', '.join(available)})"
            if available
            else " (the mesh carries no data arrays)"
        )
        raise ValueError(
            f"meshio++: no point_data or cell_data array named {name!r}{detail}"
        )

    values = []
    if is_point:
        # Point data lives on the drawn mesh: a boundary extraction carries
        # point_data through its point compaction, so the drawn mesh's node ids
        # index the already-subset array.
        if name not in draw.point_data:
            raise ValueError(
                f"meshio++: point_data array {name!r} did not survive boundary "
                "extraction"
            )
        arr = np.asarray(draw.point_data[name])
        ncomp = _num_components(arr, len(draw.points))
        flat = arr.reshape(-1)
        for face in faces:
            # Mean of the corner values, summed in node order then divided
            # once -- the same shape as the projection's depth accumulation.
            nodes = face.nodes
            total = 0.0
            count = 0
            for p in nodes:
                total += _scalarize(flat, int(p), ncomp, spec.component)
                count += 1
            values.append(float("nan") if count == 0 else total / float(count))
    else:
        cell_values = _flatten_cell_data(source, name, spec.component)
        parents = _flatten_parent_ids(draw)
        for face in faces:
            cell = face.source_cell
            if parents:
                if cell < 0 or cell >= len(parents):
                    values.append(float("nan"))
                    continue
                cell = parents[cell]
            if cell < 0 or cell >= len(cell_values):
                values.append(float("nan"))
            else:
                values.append(cell_values[cell])

    # Auto range over the DRAWN faces, not the whole array: the visible figure
    # then spans the whole colorbar. (ParaView's default is the whole array;
    # this is documented as differing.) Non-finite values are excluded, with
    # FiniteStats' seed-on-first-finite rule rather than +/-inf sentinels.
    lo = 0.0
    hi = 0.0
    seen = False
    for v in values:
        if not math.isfinite(v):
            continue
        if not seen:
            lo = v
            hi = v
            seen = True
        else:
            if v < lo:
                lo = v
            if v > hi:
                hi = v
    vmin = lo if spec.vmin is None else float(spec.vmin)
    vmax = hi if spec.vmax is None else float(spec.vmax)
    if vmin > vmax:
        raise ValueError("meshio++: vmin must not exceed vmax")
    return FaceColors(True, values, vmin, vmax, table)
